python/uv/private/uv.bzl

```diff
@@ -18,36 +18,366 @@ EXPERIMENTAL: This is experimental and may be removed without notice
+
+    Args:
+        module_ctx: the context.
+        uv_repository: the rule to create a uv_repository override.
+
+    Returns:
+        A dictionary for each version of the `uv` to configure.
+    """
+    defaults = {}
     for mod in module_ctx.modules:
-        for toolchain in mod.tags.toolchain:
-            if not mod.is_root:
-                fail(
-                    "Only the root module may configure the uv toolchain.",
-                    "This prevents conflicting registrations with any other modules.",
-                    "NOTE: We may wish to enforce a policy where toolchain configuration is only allowed in the root module, or in rules_python. See https://github.com/bazelbuild/bazel/discussions/22024",
-                )
-
-            uv_repositories(
-                uv_version = toolchain.uv_version,
-                register_toolchains = False,
+        for default_attr in mod.tags.default:
+            _configure(
+                defaults,
+                version = default_attr.version,
+                base_url = default_attr.base_url,
+                manifest_filename = default_attr.manifest_filename,
+                platform = default_attr.platform,
+                compatible_with = default_attr.compatible_with,
+                target_settings = default_attr.target_settings,
+            )
+
+    versions = {}
+    for mod in module_ctx.modules:
+        last_version = None
+        for config_attr in mod.tags.configure:
+            last_version = config_attr.version or last_version or defaults["version"]
+            if not last_version:
+                fail("version must be specified")
+
+            specific_config = versions.setdefault(
+                last_version,
+                {
+                    "base_url": defaults.get("base_url", ""),
+                    "manifest_filename": defaults["manifest_filename"],
+                    "platforms": dict(defaults["platforms"]),  # copy
+                },
+            )
+
+            _configure(
+                specific_config,
+                base_url = config_attr.base_url,
+                manifest_filename = config_attr.manifest_filename,
+                platform = config_attr.platform,
+                compatible_with = config_attr.compatible_with,
+                target_settings = config_attr.target_settings,
+                sha256 = config_attr.sha256,
+                urls = config_attr.urls,
             )
 
+    versions = {
+        version: config
+        for version, config in versions.items()
+        if config["platforms"]
+    }
+    if not versions:
+        return struct(
+            names = ["none"],
+            labels = {
+                # NOTE @aignas 2025-02-24: the label to the toolchain can be anything
+                "none": str(Label("//python:none")),
+            },
+            compatible_with = {
+                "none": ["@platforms//:incompatible"],
+            },
+            target_settings = {},
+        )
```

Comment:
Ah, this is clever. Basically causing a no-op, compatible-with-nothing toolchain to be generated?

Reply:
Correct. I think this is one of the best discoveries in this design.

Reply:
Maybe we should just have the toolchain target defined next to python:none? Like //python:incompatible_toolchain.

Reply:
I think the problem with my mental modelling of the design is that we need to create a hub repo with something so that `register_toolchains(hub:all)` in rules_python still worksno matter what. So that is why this is needed.